### ISSUE TITLE
Scoping the adoption of this library

### ISSUE THREAD

**Issue by nicolabortignon** [codesuki/react-d3-components#57]
Hi there,
first of all, thanks for the great work done so far.

I'm migrating my frontend from angular to react. I have couple of graphs that are done via d3.
I was really interest to adopting this library as middleware between react and d3.

The graph that I would like to 'port' are pretty specific, the first is an audio waveform, the other is a sunburst (http://bl.ocks.org/kerryrodden/7090426) 

Do you think I can try to integrate the two graphs into the library via PR?
How complicated is to convert a standalone d3 graph into a react-d3-component ?

**Comment by codesuki** [comment]
Hey there! 
I would be very happy if you would send a pull request for the graphs! 
Personally I might do some audio related stuff in the near future so it would be very interesting for me, too.

The integration shouldn't be too hard I think. I also used the `bl.ocks.org` page as a guide to implement the current graphs.
I checked the sunburst example and it's using `d3.svg.arc()` which is also used by the PieChart. 
https://github.com/codesuki/react-d3-components/blob/master/src/PieChart.jsx
So basically instead of letting D3 create DOM elements you get the needed parameters from `d3.svg.arc` etc and create your own DOM elements with React.

For the mouse over effects just set the right props/state for the sub components.

The audio waveform seems very simple judging from `http://bl.ocks.org/jm3/9870108`. 
Could be possible to put some animations on that too, in the near future.

If you need help just write here or in the pullreq. and we can discuss.

**Comment by nicolabortignon** [comment]
Hi Neri, thank you for the answer! 

So i put together the example for the waveform in d3. you can have a look [here](http://htmlpreview.github.io/?https://raw.githubusercontent.com/nicolabortignon/D3js-Audio-Waveform/master/index.html) and the repo is [here](https://github.com/nicolabortignon/D3js-Audio-Waveform)

Eventually I'd add a event listener when clicking any of the bar.

There are couple of things that I'm not 100% sure how can be implemented.
As you can see, there are 4 graphs each with a different sample rate that results in a different visual density. On my 24inch screen (1920x1080) the graph with 500 samples has the right graphical effect.
On my smartphone the one with 100 looks the best. On a retina it is probably the one with 1000 samples  (as a rule of thumb, each bar is 1 pixel, each space 1 px, so the closest the space available is to 2*number of sample and the best effect we have).

I need to listen to the page resize event, in order to change the sampling rate. I have no idea how to do that with react. thoughts? 

I'll start a PR and try to get a component out of this.

**Comment by nicolabortignon** [comment]
Also, would be good if you can give me some pointers how to develop the PR.
I just forked the repo, but didn't find any build command.

**Comment by codesuki** [comment]
To build: npm run build:js or just webpack.
I'll answer the message above after I arrived at work.

**Comment by codesuki** [comment]
Personally I have an alias for `webpack && cp dist/react-d3-components.js example` 
And then I just refresh the example page.. could enable auto-rebuild but I don't build that often though.

About the resizing does https://github.com/codesuki/react-d3-components/issues/9 help you?

**Comment by nicolabortignon** [comment]
Hi! and thank you for the answer. Now I can build and test with no problem.
I'm working on a fork, and i just noticed that when i start having a high amount of bars in the graph the rollover look a bit laggish (especially on mid power machines).

Have a look, the second graph is suppose to be the waveform https://htmlpreview.github.io/?https://raw.githubusercontent.com/nicolabortignon/react-d3-components/master/example/index.html

Is that normal ?

**Comment by codesuki** [comment]
Hey nice graph!
Do you mean that the highlighted bar is kinda 'late' ? 
I noticed that recently with the normal tooltip, too. (I remember that I kinda had that fixed in some earlier version though)

I am pretty sure it's fixable. Maybe the setState in React is rendered a little bit late? 
Have to investigate that more.

**Comment by nicolabortignon** [comment]
For now I just removed the listener and suddenly is super fast! 
I'll keep it like this for now.

Also, I did the mistakes of pull requesting directly from my master branch. I added new features, and I'm getting closer to what I wanted, but I'm stuck with the pull request -1: (git newbie)

**Comment by codesuki** [comment]
Do you have a demo of the super fast version? Because I don't really know what part you thought was slow. 

Not such a huge problem to do the pull request from master, although maybe not usual.
You could to `git checkout -b feature/waveform` and it will have the current state of your master.
Then commit there and do a pull request from there. Then you can reset your master. But I don't mind actually.

I'll look at the questions in your pull request when I have some free minutes!

**Comment by codesuki** [comment]
I would say just leave the pull request like this without starting a new branch, that's the simplest. :)

**Comment by nicolabortignon** [comment]
This page automatically fetch the current head committed on github, https://htmlpreview.github.io/?https://raw.githubusercontent.com/nicolabortignon/react-d3-components/master/example/index.html

it is already the fast version. Regarding the question on the pull request, few of them has already been addressed (I updated the post), and the commit is in my local repo post PR.

**Comment by codesuki** [comment]
Oh, so maybe I never actually saw the sluggish version! Since now it's the same :)

**Comment by nicolabortignon** [comment]
Yeah, probably. Before just to give you an idea was taking ~2 seconds to
highlight the bar, and when fast moving between bars only one every 20 was
highlighted. The event stack was reporting a sequence of loop. I restarted
from 0 and the problem was gone.
Pretty sure my fault.

The version on the pending PR seams stable, as well as the one in my local
repo

**Comment by codesuki** [comment]
I see, glad it's fixed!

I like how you made the graph responsive with the viewBox property! That will come in handy in all the other graphs too! Good work :)

Now only the mouse events are missing I guess?

**Comment by nicolabortignon** [comment]
Thanks! 
Yep, I need to attach the listeners, and figuring out why with more than 800 data point it fails to render.

**Comment by codesuki** [comment]
I will have a look at the 800 point thing too. If I find something I'll let you know.
